### PR TITLE
allow engine/name support in volumes

### PR DESCRIPTION
**Pull request by vieux**
Fix https://github.com/docker/swarm/issues/1302

### CODE REVIEW

**Comment by vieux** [comment]
ping @mbentley can you please try this one ?

**Comment by mbentley** [comment]
Looks great!  Thanks @vieux!

```
$ docker -H tcp://127.0.0.1:3375 info
Containers: 1
Images: 62
Role: primary
Strategy: spread
Filters: health, port, dependency, affinity, constraint
Nodes: 1
 localdev: 192.168.56.4:2375
  └ Status: Healthy
  └ Containers: 1
  └ Reserved CPUs: 0 / 1
  └ Reserved Memory: 0 B / 2.061 GiB
  └ Labels: executiondriver=native-0.2, kernelversion=3.16.0-4-amd64, operatingsystem=Debian GNU/Linux stretch/sid (containerized), storagedriver=aufs
  └ Error: (none)
  └ UpdatedAt: 2016-01-22T02:30:26Z
CPUs: 1
Total Memory: 2.061 GiB
Name: 34f77f63050b
$ docker -H tcp://127.0.0.1:3375 volume ls
DRIVER              VOLUME NAME
local               localdev/haproxy
local               localdev/aa7d97f5ff6f4f60d64f5e19b3ca573938a27d5a803bea5e3560ce1a50f8c305
```

**Comment by vieux** [comment]
@abronan @dongluochen failure unrelated, I'm going to fix that in another version.

Please review

File: api/handlers.go [hunk 1]
```diff
@@ -808,9 +814,9 @@ func proxyNetwork(c *context, w http.ResponseWriter, r *http.Request) {
 // Proxy a request to the right node
 func proxyVolume(c *context, w http.ResponseWriter, r *http.Request) {
 	var name = mux.Vars(r)["volumename"]
-	if volume := c.cluster.Volume(name); volume != nil {
-		err := proxy(c.tlsConfig, volume.Engine.Addr, w, r)
-		volume.Engine.CheckConnectionErr(err)
+	if volume := c.cluster.Volumes().Get(name); volume != nil {
```
Comment (dongluochen) [inline api/handlers.go]: Do you want to update function name `proxyVolume` here?

Comment (vieux) [inline api/handlers.go]: yes I'll also update `proxyNetwork` at the same time, it's not a proxy anymore

File: cluster/volume.go [hunk 2]
```diff
@@ -8,3 +8,42 @@ type Volume struct {
...
+// Get returns a volume using it's ID or Name
+func (volumes Volumes) Get(name string) *Volume {
+	// Abort immediately if the name is empty.
+	if len(name) == 0 {
+		return nil
+	}
+
+	candidates := []*Volume{}
+
+	// Match name, /name or engine/name.
```
Comment (dongluochen) [inline cluster/volume.go]: The comment doesn't match code. The logic of this function is a little confusing to me. For example why 'size >1' is not a valid case. I'd suggest add comment to clarify.

File: cluster/volume.go [hunk 3]
```diff
@@ -8,3 +8,42 @@ type Volume struct {
...
+		}
+	}
+
+	if size := len(candidates); size == 1 {
+		return candidates[0]
+	} else if size > 1 {
+		return nil
+	}
+
+	// Match name, /name or engine/name.
```
Comment (dongluochen) [inline cluster/volume.go]: Comment doesn't match code.

**Comment by abronan** [comment]
LGTM

**Comment by dongluochen** [comment]
LGTM.

**Pull request merged**